docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1119 @@
+  necessary.
+- Be strictly correct when it comes to data integrity. Occasional failures of availability
+  are tolerable, occasional data loss is not.
+
+## Non Goals
+
+The changes in this RFC intentionally isolate the design decision of how to define
+logical generations IDs and object storage format in a way that is somewhat flexible with
+respect to how actual orchestration of failover works.
+
+This RFC intentionally does not cover:
+
+- Failure detection
+- Orchestration of failover
+- Standby modes to keep data ready for fast migration
+- Intentional multi-writer operation on tenants (multi-writer scenarios are assumed to be transient split-brain situations).
+- Sharding.
+
+## Impacted Components
+
+pageserver, control plane, safekeeper (optional)
+
+## Implementation Part 1: Correctness
+
+### Summary
+
+- **Generation numbers** are introduced for pageserver node lifetimes and tenant attachments.
+
+  - node generation increments each time a pageserver starts, before it can write to S3
+  - attachment generation increments each time the control plane modifies a tenant (`Project`)'s assigned pageserver (an _attachment_'s lifetime is the association between a tenant and a pageserver)
+  - the two generations are independent: a pageserver may restart while keeping the same
+    attachment generation numbers, and attachments may be changed while a pageserver's
+    generation number stays the same.
+
+- **Object keys are suffixed** with a tuple of attachment & node generation numbers
+- **Safety in split brain for multiple nodes running with
+  the same node ID** is provided by the pageserver node generation in the object key: the concurrent nodes
+  will not write to the same key.
+- **Safety for multiply-attached tenants** is provided by the
+  tenant attach generation in the object key: the competing pageservers will not
+  try to write to the same keys.
+- **Safety for deletions** is provided by pageservers validating with control plane that neither pageserver process nor the attachment has been superseded since it last updated a timeline's index.
+- **The control plane is used to issue generation numbers** to avoid the need for
+  a built-in consensus system in the pageserver, although this could in principle
+  be changed without changing the storage format.
+
+### Generation numbers
+
+Two logical entities will get new "generation numbers", which are monotonically increasing
+integers with a global guarantee that they will not be re-used:
+
+- Pageserver processes: for a given pageserver node ID, each time a pageserver process starts,
+  the per-node-ID generation number increases.
+- Tenant attachments: for a given tenant, each time its attachment is changed, a per-tenant generation
+  number increases.
+
+This provides some important invariants:
+
+- If there are multiple pageservers running with the same node ID, they are guaranteed to have
+  a different generation number.
```

Comment:
Different "node generation number"? Or different "attachment generation number"? Or does just "generation number" refer to the combination of the two?

Reply:
In this context it meant node generation number, but this text is gone now.